lib/python/bind_data_array.h

```diff
@@ -64,6 +64,65 @@ void bind_common_mutable_view_operators(pybind11::class_<T, Ignored...> &view) {
       });
 }
 
+template <class D> auto cast_to_dict_key(const py::handle &obj) {
+  using key_type = typename D::key_type;
+  if constexpr (std::is_same_v<key_type, std::string>) {
+    return obj.cast<std::string>();
+  } else {
+    return key_type{obj.cast<std::string>()};
+  }
+}
+
+template <class D> auto cast_to_dict_value(const py::handle &obj) {
+  using val_type = typename D::mapped_type;
+  return obj.cast<val_type>();
+}
+
+template <class T, class... Ignored, class Set>
+void bind_dict_update(pybind11::class_<T, Ignored...> &view, Set &&set_item) {
+  view.def(
+      "update",
+      [set_item](T &self, const py::object &other, const py::kwargs &kwargs) {
+        if (!other.is_none()) {
+          if (py::hasattr(other, "keys")) {
+            // other is dict-like
+            for (const auto &key : other.attr("keys")()) {
+              set_item(self, cast_to_dict_key<T>(key),
+                       cast_to_dict_value<T>(other[key]));
+            }
+          } else {
+            // other is sequence of tuples
+            for (const auto &item : other) {
+              const auto &t = item.cast<py::tuple>();
+              set_item(self, cast_to_dict_key<T>(t[0]),
+                       cast_to_dict_value<T>(t[1]));
+            }
+          }
+        }
+
+        for (const auto &item : kwargs) {
+          set_item(self, cast_to_dict_key<T>(item.first),
+                   cast_to_dict_value<T>(item.second));
+        }
+      },
```

Comment:
Would it be easier to simply handle a `dict` on the C++ side, and do the translation of tuple and kwargs in Python?

Reply:
This gave me an idea. I used `py::dict` to handle the arguments and then copy from that into our `Dict`. Moving it to Python would only save a few characters in the implementation...